In [74]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [75]:
# (Class 1 Activity 4)

# Read the CSV file into a Pandas DataFrame
lending_data_df = pd.read_csv(Path('lending_data.csv'))

# Display sample data
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [76]:
# Separate the data into labels and features (Class 1 Activity 4)

# Separate the y variable, the labels
y = lending_data_df['loan_status']

# Separate the X variable, the features
X = lending_data_df.drop(columns = ['loan_status'])

In [77]:
# Review the y variable Series (Class 1 Activity 6)
display(y.head())
display(y.tail())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, dtype: int64

The above output is consistent with the csv file

In [88]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

The above output is consistent with the csv file

In [89]:
# Check the balance of our target values
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

75036 + 2500 = 77536 which is consistent with the number of records in the csv file.

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [79]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

#(Class 2 Activity 8)
# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    random_state = 1,
    stratify = y)

In [80]:
X_train.shape 
print('Training dataset:', X_train.shape[0], 'rows &', X_train.shape[1], 'columns')

X_test.shape 
print('Test dataset:', X_test.shape[0], 'rows &', X_test.shape[1], 'columns')

Training dataset: 58152 rows & 7 columns
Test dataset: 19384 rows & 7 columns


Total of Training dataset + Test dataset is 77536 which is consistent with the total number of rows in the csv file

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [81]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Class 1 Activity 06
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(
    solver ='lbfgs',
    max_iter = 200,
    random_state = 1)

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(max_iter=200, random_state=1)

In [82]:
print(f'Training Data Score: {round(classifier.score(X_train, y_train), 3)}')
print(f'Testing Data Score: {round(classifier.score(X_test, y_test), 3)}')

Training Data Score: 0.991
Testing Data Score: 0.992


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [83]:
# Class 2 Activities 6 & 8
# Make a prediction using the testing data
 
y_pred = classifier.predict(X_test)
pd.DataFrame({'Prediction': y_pred, 'Actual': y_test}).reset_index(drop = True).head()

Prediction  Actual
0           0       0
1           0       1
2           0       0
3           0       0
4           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [84]:
# Class 2 Activity 6
# Generate a confusion matrix for the model
confusion_matrix(y_test, y_pred)

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [85]:
print(confusion_matrix(y_test, y_pred))

[[18679    80]
 [   67   558]]


In [91]:
# Print the classification report for the model (Class 2 Activity 8)
testing_report = classification_report(y_test, y_pred)
print(f'The classification report with the original data shows:\n{testing_report}')

The classification report with the original data shows:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 


After observing the above results, the accuracy seems to be good enough to start exploring this kind of algorithms to assess loan risk and to identify the creditworthiness of borrowers.

For the 0 (healthy loans) label: Precision, Recall and f1-score are all 1.00.

This means that the model has high precision and does really well when predicting a healthy loan (except for a small percentage of false positives).
The model also has high recall which means that it can identify most of the healthy loans (small percentage of false negatives). 
The F1-score is also high, indicating a good balance between precision and recall.

For high-risk loans the precision is 0.87, the recall is 0.89, and the f1-score is 0.88, meaning that this model does moderately well in predicting instances.
_______________________________________________________________________

For the 1 (high-risk loans) label: Precision = 0.87, Recall = 0.89, f1-score = 0.88.

The model's performance for high-risk loans is relatively good but not as perfect as for healthy loans. It has a precision of 0.87 indicating that when the model predicts a high-risk loan, it is correct about 87% of the time, but there would still be some false positives. The recall is 0.89, suggesting that the model captures 89% of the actual high-risk loans but may still miss a few (false negatives). The F1-score is 0.88, indicating a reasonable balance between precision and recall.

In summary, this logistic regression model performs very well for both the 0 (healthy loan) and 1 (high-risk loan) labels, with high accuracy, precision, recall, and F1-scores for both classes. 
The data has a  disproportionately larger number of healthy loans compared to high-risk loans so it is especially good at correctly identifying healthy loans, while its performance for high-risk loans is not quite as strong.

---

In [87]:
# from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# # Initialize the logistic regression model
# model = LogisticRegression(random_state=1)

# # Train the model on the training data
# model.fit(X_train, y_train)

# # Make predictions on the test set
# y_pred = model.predict(X_test)

# # Calculate accuracy
# accuracy = accuracy_score(y_test, y_pred)
# print(f'Accuracy: {accuracy:.2%}')

# # Calculate precision
# precision = precision_score(y_test, y_pred)
# print(f'Precision: {precision:.2%}')

# # Calculate recall
# recall = recall_score(y_test, y_pred)
# print(f'Recall: {recall:.2%}')

# # Calculate F1 score
# f1 = f1_score(y_test, y_pred)
# print(f'F1 Score: {f1:.2%}')


Accuracy: 99.24%
Precision: 87.46%
Recall: 89.28%
F1 Score: 88.36%
